### PR TITLE
Add Token (Smart Contract) feature

### PR DESCRIPTION
**Pull request by SheldonYS**
# PR Details

1. add SPoS related TX
2. add Smart Contract (Token) related TX

## Description

Add the following transaction types support:
#### SPoS
- [X] Contend Slot
- [X] Release Slot

#### Smart Contract
- [X] Register Contract
- [X] Execute Contract

## Related Issue

None, new features.

## Motivation and Context

add support to Mainchain new feature

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by ncying** [comment]
@zl730 please help to double check the code

**Comment by zl730** [comment]
Please try to use "new" features of java 8 as much as possible.
Here's the [tutorial](https://howtodoinjava.com/java-8-tutorial/)

**Comment by zl730** [comment]
And it's actually very rear to see a long last java project without lombok.
A [video tutorial](https://projectlombok.org/) for you

File: src/v/systems/contract/DataEntry.java [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        return data;
+    }
+
+    public void setData(byte[] data) {
+        this.data = data;
+    }
+
+    public static DataEntry publicKey(String publicKey) {
+        DataEntry dataEntry = new DataEntry();
+        dataEntry.setType(DataType.PublicKey);
```
Comment (zl730) [inline src/v/systems/contract/DataEntry.java]: why not use constructor? or builder?

File: src/v/systems/contract/Contract.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        textual = new ContractTextual();
+    }
+
+    @Override
+    public byte[] toBytes() throws SerializationError {
+        try {
+            byte[] triggerBytes = serializeField(triggers);
+            byte[] descriptorBytes = serializeField(descriptors);
+            byte[] stateVarBytes = serializeField(stateVariables);
+            byte[][] result = new byte[][]{
```
Comment (zl730) [inline src/v/systems/contract/Contract.java]: Why not use `Bytes` in Guava, or `ByteBuffer`?

File: src/v/systems/utils/BytesHelper.java [hunk 3]
```diff
@@ -1,9 +1,8 @@
 package v.systems.utils;
 
-import org.bitcoinj.core.Base58;
-
 import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
 import java.util.List;
 
 public class BytesHelper {
```
Comment (zl730) [inline src/v/systems/utils/BytesHelper.java]: methods here are not thread-safe. please use guava instead.

File: src/v/systems/contract/Contract.java (hunk 2 above)
Comment (SheldonYS) [reply src/v/systems/contract/Contract.java]: `Guava` is great library. However, we only use convert integer to bytes and concat functions from it. `BytesHelper.java`(which only 80 Lines) can meet the requirement. I'll consider using `Guava` if we need more Guava function later. For this version, I just want to keep SDK in lightweight.

**Pull request merged**